lib/resque/scheduler/scheduling_extensions.rb

```diff
@@ -149,6 +123,125 @@ def remove_schedule(name)
+            set_schedule(name, job_spec)
+          end
+        end
+
+        # ensure only return the successfully saved data!
+        reload_schedule!
+      end
+
+      def atomic_setup_schedule(schedule_hash)
+        schedule = {}
+
+        Resque.redis.evalsha(
+          startup_sha,
+          keys: [:schedules, :persisted_schedules, :schedules_changed],
+          argv: prepare_lua_schedule(schedule_hash)
+        ).each_slice(2) do |name, config|
+          schedule[name] = decode(config)
+        end
+
+        @schedule = schedule
+      end
+
+      def startup_sha(refresh = false)
+        @startup_sha = nil if refresh
+
+        @startup_sha ||=
+          Resque.redis.script(:load, <<-EOF.gsub(/^ {14}/, ''))
+            local schedules_key = KEYS[1]
+            local persisted_schedules_key = KEYS[2]
+            local changed_schedules_key = KEYS[3]
+
+            local remove_schedule = function(schedule)
+              redis.call("HDEL", schedules_key, schedule)
+              redis.call("SREM", persisted_schedules_key, schedule)
+              redis.call("SADD", changed_schedules_key, schedule)
+            end
+
+            local get_non_persistent_schedules = function()
+              local all_schedules = redis.call("HKEYS", schedules_key)
+              local non_persistent_schedules = {}
+              for _,schedule in ipairs(all_schedules) do
+                local schedule_is_persisted = \
+                  redis.call("SISMEMBER", persisted_schedules_key, schedule)
+                if schedule_is_persisted == 0 then
+                  table.insert(non_persistent_schedules, schedule)
+                end
+              end
+              return non_persistent_schedules
+            end
+
+            local set_schedule = function(name, config, persisted)
+              redis.call("HSET", schedules_key, name, config)
+              redis.call("SADD", changed_schedules_key, name)
+              if persisted == "true" then
+                redis.call("SADD", persisted_schedules_key, name, config)
+              end
+            end
+
+            local non_persistent_schedules = get_non_persistent_schedules()
+
+            for _,schedule in ipairs(non_persistent_schedules) do
+              remove_schedule(schedule)
+            end
+
+            -- i:   schedule_name
+            -- i+1: encoded_schedule_config
+            -- i+2: is_schedule_persisted?
+            for index=1,#ARGV,3 do
+              set_schedule(ARGV[index], ARGV[index+1], ARGV[index+2])
+            end
+
+            return redis.call("HGETALL", schedules_key)
+          EOF
+      end
+
+      def prepare_lua_schedule(schedule_hash)
+        prepared_schedule = prepare_schedule(schedule_hash)
+        res = []
+        prepared_schedule.each do |name, config|
+          persist = config.delete(:persist) || config.delete('persist')
+          res << name << encode(config) << !persist.nil?
+        end
+        res
+      end
```

Comment:
`prepare_schedules(schedule_hash).map { |name, config| [name, config.delete(:persist), encode(config)] }.flatten` is simpler.. but this is just me looking for an opportunity to golf. That said, `#map` would be nice to use here as it communicates an intent to transform. Note that the ordering changed in the array, but if it was implemented this way you'd have to as it'd be evaluated left-to-right.